File: pylimerpredictor/predictors/mmt_predictor.py
#!/usr/bin/env python

import warnings

from pylimer_tools.calc.miller_macosko_theory import (
    compute_modulus_decomposition,
    compute_weight_fraction_of_backbone,
    compute_weight_fraction_of_dangling_chains,
    compute_weight_fraction_of_soluble_material,
)
from quantityfield.units import ureg

from pylimerpredictor.models.prediction_input import PredictionInput

"""
Shortcut to acquire all data from MMT given some parameters.
"""


def predict_mmt_results(prediction_input: PredictionInput) -> dict:
    """
    Compute MMT data for given parameters.
    """
    r = prediction_input.stoichiometric_imbalance
    p = prediction_input.crosslink_conversion
    f = prediction_input.crosslink_functionality
    ge_1 = prediction_input.plateau_modulus
    temperature = prediction_input.temperature
    density = prediction_input.density
    Mw1 = prediction_input.n_beads_monofunctional * prediction_input.bead_mass
    Mw2 = prediction_input.n_beads_bifunctional * prediction_input.bead_mass
    MwX = prediction_input.n_beads_xlinks * prediction_input.bead_mass
    b2 = prediction_input.get_b2()

    n_chains = 1e7
    n_chains_mono = int((2 * n_chains * (1 - b2)) / b2)
    # require Mw1, Mw2, MwX to have the same units
    if (
        not isinstance(Mw1, ureg.Quantity)
        or not isinstance(Mw2, ureg.Quantity)
        or not isinstance(MwX, ureg.Quantity)
    ):
        raise ValueError("All Mw parameters must have the same units")

    if Mw1.check("[mass]/[substance]"):
        Na = 6.022e23 * ureg("1/mol")
        Mw1 = (Mw1 / Na).to("kg")
        Mw2 = (Mw2 / Na).to("kg")
        MwX = (MwX / Na).to("kg")

    assert Mw1.check("[mass]") and Mw2.check("[mass]") and MwX.check("[mass]")

    Mw_eff = (n_chains * Mw2 + n_chains_mono * Mw1) / (n_chains + n_chains_mono)
    data = {
        "Mw_eff [kg]": Mw_eff.to("kg").magnitude,
        "MMT nu [nm^-3]": (density / (Mw_eff)).to("nm^-3").magnitude,
    }
    try:
        G_MMT_phantom, G_MMT_entanglement, g_anm, g_pnm = compute_modulus_decomposition(
            network=None,
            crosslinker_type=2,
            r=r,
            p=p,
            f=f,
            g_e_1=ge_1,
            temperature=temperature,
            nu=(density / (Mw_eff)),
            ureg=ureg,
            b2=b2,
        )
    except ValueError as e:
        warnings.warn(f"Error computing modulus decomposition: {e}")
        return {}
    data["g_phantom"] = G_MMT_phantom.to("MPa").magnitude
    # TODO: this correction for b2 must be reconsidered
    b2_entanglement_correction = (
        ((Mw2 * n_chains) / (Mw1 * n_chains_mono + Mw2 * n_chains))
        .to("dimensionless")
        .magnitude
    )
    data["Mw b2 entanglement correction"] = b2_entanglement_correction
    data["g_entangled"] = (
        G_MMT_entanglement.to("MPa").magnitude * b2_entanglement_correction
    )
    data["g_eq"] = data["g_entangled"] + data["g_phantom"]
    data["g_anm"] = g_anm.to("MPa").magnitude
    data["g_pnm"] = g_pnm.to("MPa").magnitude
    functionality_per_type = {1: 2, 2: f, 4: 1}
    n_crosslinks = int(((n_chains * 2 + n_chains_mono) * r) / (f))
    total_weight = n_chains_mono * Mw1 + n_chains * Mw2 + n_crosslinks * MwX
    weight_fractions = {
        1: (n_chains * Mw2) / (total_weight),
        2: (n_crosslinks * MwX) / (total_weight),
        4: (n_chains_mono * Mw1) / (total_weight),
    }

    for k in weight_fractions.keys():
        data[f"MMT weight fraction of type {k}"] = weight_fractions[k]

    data["w_soluble"] = compute_weight_fraction_of_soluble_material(
        network=None,
        crosslinker_type=2,
        functionality_per_type=functionality_per_type,
        weight_fractions=weight_fractions,
        r=r,
        p=p,
        b2=b2,
    )
    data["w_backbone"] = compute_weight_fraction_of_backbone(
        network=None,
        crosslinker_type=2,
        functionality_per_type=functionality_per_type,
        weight_fractions=weight_fractions,
        r=r,
        p=p,
        b2=b2,
    )
    data["w_dangling"] = compute_weight_fraction_of_dangling_chains(
        network=None,
        crosslinker_type=2,
        functionality_per_type=functionality_per_type,
        weight_fractions=weight_fractions,
        r=r,
        p=p,
        b2=b2,
    )

    return data

File: src/worker_predictor/mmt/mmt_predictor.ts
import type { PredictionInput } from "../../entity/PredictionInput";
import { ModulusPredictionOutput } from "../../entity/PredictionOutput";
import type { Predictor } from "../predictor";
import Qty from "js-quantities";
import {
  computeMillerMacoskoProbabilities,
  computeModulusDecomposition,
  computeWeightFractionDanglingChains,
  computeWeightFractionBackbone,
  computeWeightFractionSolubleMaterial,
} from "./mmt_calculations";

export class MMTPredictor implements Predictor {
  async predict(input: PredictionInput): Promise<ModulusPredictionOutput> {
    console.log("Predicting MMT results...", input);

    try {
      const result = this.predictMMTResults(input);
      return result;
    } catch (error) {
      throw new Error(
        `Failed to predict MMT results: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  private predictMMTResults(input: PredictionInput): ModulusPredictionOutput {
    const r = input.stoichiometric_imbalance;
    const p = input.crosslink_conversion;
    const f = input.crosslink_functionality;
    const ge1 = input.plateau_modulus;
    const temperature = input.temperature;
    const density = input.density;
    const b2 = input.b2;

    // Calculate molecular weights
    const Mw1 = input.bead_mass.mul(input.n_beads_monofunctional);
    const Mw2 = input.bead_mass.mul(input.n_beads_bifunctional);
    const MwX = input.bead_mass.mul(input.n_beads_xlinks);

    // Convert molecular weights from g/mol to kg (actual mass) using Avogadro's number
    const Na = Qty(6.022e23, "1/mol");
    const Mw1Kg = Mw1.div(Na).to("kg");
    const Mw2Kg = Mw2.div(Na).to("kg");
    const MwXKg = MwX.div(Na).to("kg");

    // Calculate number of chains
    const nChains = 1e7;
    const nChainsMono = Math.trunc((2 * nChains * (1 - b2)) / b2);

    // Calculate effective molecular weight
    const MwEff = Mw2Kg.mul(nChains)
      .add(Mw1Kg.mul(nChainsMono))
      .div(nChains + nChainsMono);

    // Calculate strand number density
    const nu = density.div(MwEff);

    // Compute modulus decomposition
    const [gMmtPhantom, gMmtEntanglement, gAnm, gPnm] = computeModulusDecomposition(
      r,
      p,
      f,
      nu,
      temperature,
      ge1,
      b2
    );

    // Calculate b2 entanglement correction
    const b2EntanglementCorrection = Mw2.mul(nChains)
      .div(Mw1.mul(nChainsMono).add(Mw2.mul(nChains)))
      .scalar;

    // Apply correction to entanglement modulus
    const correctedEntanglementModulus = gMmtEntanglement.mul(b2EntanglementCorrection);

    // Calculate weight fractions for MMT
    const nCrosslinks = Math.trunc(((nChains * 2 + nChainsMono) * r) / f);
    const totalWeight = Mw1.mul(nChainsMono)
      .add(Mw2.mul(nChains))
      .add(MwX.mul(nCrosslinks));

    const weightFractions: Record<number, number> = {
      1: Mw2.mul(nChains).div(totalWeight).scalar, // bifunctional
      2: MwX.mul(nCrosslinks).div(totalWeight).scalar, // crosslinkers
      4: Mw1.mul(nChainsMono).div(totalWeight).scalar, // monofunctional
    };

    const functionalityPerType: Record<number, number> = {
      1: 2, // bifunctional chains
      2: f, // crosslinkers
      4: 1, // monofunctional chains
    };

    // Compute probabilities
    const [alpha, beta] = computeMillerMacoskoProbabilities(r, p, f, b2);

    // Compute weight fractions
    const wSoluble = computeWeightFractionSolubleMaterial(
      functionalityPerType,
      weightFractions,
      alpha,
      beta
    );

    const wDangling = computeWeightFractionDanglingChains(
      functionalityPerType,
      weightFractions,
      alpha,
      beta
    );

    // Create output
    const output = new ModulusPredictionOutput();
    output.phantom_modulus = gMmtPhantom.to("MPa");
    output.entanglement_modulus = correctedEntanglementModulus.to("MPa");
    output.w_soluble = Qty(wSoluble, "");
    output.w_dangling = Qty(wDangling, "");

    return output;
  }
}
